File: code-tiny/tools/struts/xml_utils.py
from __future__ import annotations

import os
import xml.etree.ElementTree as ET
from pathlib import Path
from typing import Dict, Iterable, Optional

from tools.struts.models import Diagnostic, SourceSpan


def parse_xml(root: str, file_path: str, code_prefix: str) -> tuple[Optional[ET.Element], SourceSpan, tuple[Diagnostic, ...]]:
    project_root = Path(root).resolve()
    candidate = Path(file_path)
    absolute = candidate.resolve() if candidate.is_absolute() else (project_root / candidate).resolve()
    try:
        relative = absolute.relative_to(project_root).as_posix()
    except ValueError:
        return None, SourceSpan(str(file_path)), (
            Diagnostic(f"{code_prefix}.path.outside_root", "XML path is outside the project root", "error", str(file_path)),
        )
    try:
        document = ET.parse(absolute)
    except (OSError, ET.ParseError) as exc:
        return None, SourceSpan(relative), (
            Diagnostic(f"{code_prefix}.xml.parse_error", str(exc), "error", relative),
        )
    return document.getroot(), SourceSpan(relative), ()


def local_name(tag: str) -> str:
    return tag.rsplit("}", 1)[-1]


def children(element: ET.Element, name: str) -> Iterable[ET.Element]:
    return (child for child in element if local_name(child.tag) == name)


def first_child(element: ET.Element, name: str) -> Optional[ET.Element]:
    return next(children(element, name), None)


def child_text(element: ET.Element, name: str, default: str = "") -> str:
    child = first_child(element, name)
    return text(child) if child is not None else default


def text(element: Optional[ET.Element]) -> str:
    if element is None:
        return ""
    return "".join(element.itertext()).strip()


def params(element: ET.Element) -> Dict[str, str]:
    values: Dict[str, str] = {}
    for item in children(element, "param"):
        name = (item.get("name") or "").strip()
        if name:
            values[name] = text(item)
    return values


def normalized_path(path: str) -> str:
    return os.path.normpath(path).replace(os.sep, "/")

File: rust/crates/analyzer-jvm-overlays/src/struts/xml_utils.rs
//! Port `tools/struts/xml_utils.py`.

use std::path::{Path, PathBuf};

use super::models::{Diagnostic, Params, SourceSpan};
use super::xml_dom::{parse, XmlElement};

/// `parse_xml` — resolve file dưới project root + parse; trả (root element,
/// SourceSpan(relative), diagnostics).
pub fn parse_xml(root: &str, file_path: &str, code_prefix: &str) -> (Option<XmlElement>, SourceSpan, Vec<Diagnostic>) {
    let project_root = crate::pyutil::realpath(Path::new(root));
    let candidate = PathBuf::from(file_path);
    let absolute = if candidate.is_absolute() {
        crate::pyutil::realpath(&candidate)
    } else {
        crate::pyutil::realpath(&project_root.join(&candidate))
    };
    let relative = match absolute.strip_prefix(&project_root) {
        Ok(rel) => crate::pyutil::to_posix(rel),
        Err(_) => {
            return (
                None,
                SourceSpan::from_path(file_path),
                vec![Diagnostic::new(
                    &format!("{code_prefix}.path.outside_root"),
                    "XML path is outside the project root",
                    "error",
                    file_path,
                )],
            );
        }
    };
    match std::fs::read(&absolute) {
        Err(error) => (
            None,
            SourceSpan::from_path(&relative),
            vec![Diagnostic::new(
                &format!("{code_prefix}.xml.parse_error"),
                &error.to_string(),
                "error",
                &relative,
            )],
        ),
        Ok(bytes) => match parse(&bytes) {
            Ok(document) => (Some(document), SourceSpan::from_path(&relative), Vec::new()),
            Err(message) => (
                None,
                SourceSpan::from_path(&relative),
                vec![Diagnostic::new(
                    &format!("{code_prefix}.xml.parse_error"),
                    &message,
                    "error",
                    &relative,
                )],
            ),
        },
    }
}

/// `local_name` — strip Clark `{uri}`.
pub fn local_name(tag: &str) -> String {
    tag.rsplit('}').next().unwrap_or(tag).to_string()
}

/// `children(element, name)` — direct children với local name khớp.
pub fn children<'a>(element: &'a XmlElement, name: &str) -> Vec<&'a XmlElement> {
    element
        .elements()
        .filter(|child| local_name(&child.tag) == name)
        .collect()
}

/// `first_child`.
pub fn first_child<'a>(element: &'a XmlElement, name: &str) -> Option<&'a XmlElement> {
    children(element, name).into_iter().next()
}

/// `child_text`.
pub fn child_text(element: &XmlElement, name: &str, default: &str) -> String {
    match first_child(element, name) {
        Some(child) => text(Some(child)),
        None => default.to_string(),
    }
}

/// `text(element)` — itertext().strip().
pub fn text(element: Option<&XmlElement>) -> String {
    match element {
        None => String::new(),
        Some(element) => element.itertext().trim().to_string(),
    }
}

/// `params(element)` — `<param name="...">text</param>` children.
pub fn params(element: &XmlElement) -> Params {
    let mut values = Params::new();
    for item in children(element, "param") {
        let name = item.attr("name").unwrap_or("").trim().to_string();
        if !name.is_empty() {
            values.insert(name, text(Some(item)));
        }
    }
    values
}

/// `normalized_path` (không dùng trong pipeline hiện tại).
#[allow(dead_code)]
pub fn normalized_path(path: &str) -> String {
    path.to_string()
}
